### ISSUE TITLE
Large stack size of yyParser::yy_reduce in debug builds

### ISSUE THREAD

**Issue by ignatz** [gwenn/lemon-rs#78]
Maybe you're aware of this already...

TL;DR: the sum of local variables in yy_reduce is larger than 1MB in debug builds. Which means that trivial parses will fail automatically for debug builds on some systems (e.g. Windows).

## Context

I'm using sqlite3-parser and got a report by someone building on Windows that their default stack size of 1MB isn't enough. I found this quite curious and found that it's: yyParser::yy_reduce.

```bash
$ ulimit -S -s 1024
$ cargo run --example sql_cmd "CREATE TABLE _procedure(name text primary key, type text, body text)"
...
thread 'main' has overflowed its stack
```

I looked at the generated code and was first a bit puzzled. It seems to be an instance of: https://internals.rust-lang.org/t/match-expressions-use-a-lot-of-stack-space-in-debug-mode/17173. (match expressions, ifs, ... all the same). I could "verify" this behavior by removing stack allocations in unused branches until above program would eventually run.

## Possible Solutions

 * My first instinct was to look for code-gen attributes to tell the rust compiler to always overlap local variables but didn't find anything. Hopefully, I'm just ignorant
 * Reduce local allocations, e.g. move branches into inlined functions, optimize stack size of AST types, ... :shrug: 
 * Ignore this issue

Curious to hear your thoughts

**Comment by gwenn** [comment]
See #8

**Comment by ignatz** [comment]
Awesome! I see: https://github.com/tursodatabase/libsql/issues/1051 is the same issue. Do you forsee https://github.com/gwenn/lemon-rs/issues/8 to be enough or just a piece in the puzzle?

**Comment by gwenn** [comment]
I only guess: the main difference between the C parser and Rust parser is the AST (and tagged `union` vs `enum`)...

**Comment by ignatz** [comment]
The other major difference is the compiler. Reserve space for all variables in all scopes and mutually exclusive branches in debug builds is a bespoke choice of the rust compiler

**Comment by gwenn** [comment]
Indeed, in release mode:
```sh
$ ulimit -S -s 256
[vagrant@archlinux lemon-rs]$ cargo  run -r --example sql_cmd "CREATE TABLE _procedure(name text primary key, type text, body text)"
    Finished `release` profile [optimized] target(s) in 0.03s
     Running `target/release/examples/sql_cmd 'CREATE TABLE _procedure(name text primary key, type text, body text)'`
CREATE TABLE _procedure (name text PRIMARY KEY, type text, body text);
```

**Comment by gwenn** [comment]
Could you please confirm that if we simply the actions (almost no local variables), we should reduce the stack size ?
By actions, I mean like:
https://github.com/gwenn/lemon-rs/blob/6f463674743fc5759b19747f2c07220bb7ff1663/src/parser/parse.y#L295-L299
And do you know how to monitor the max stack size used by one program (maybe dtrace) ?

**Comment by ignatz** [comment]
> Could you please confirm that if we simply the actions (almost no local variables), we should reduce the stack size ?

I don't know. We would need to test if they get elided. At the end of the day, the compiler still has to reserve space for temporaries. How, when and where is an implementation detail. Also not all temporaries could be removed because of borrow checker rules etc.

What I originally had in mind, which should work:

```rust
impl yyParser<'_> {
    fn yy_reduce(...) {
      // ...
      match yyruleno {
        // ...
        28 => {
         {
           let col = self.yy_move(-1).yy444();
           let cd = ColumnDefinition{ col_name: col.0, col_type: col.1, constraints: self.yy_move(0).yy305() };
           let mut map = IndexMap::new();
           ColumnDefinition::add_column(&mut map, cd)?;
           yylhsminor = YYMINORTYPE::yy285( map);
         }
         self[-1].minor = yylhsminor;
       }
       // ...
      }
      // ...
    }
}
```

to

```rust
impl yyParser<'_> {
    fn yy_reduce(...) {
      // ...
      match yyruleno {
        // ...
        28 => {
           self[-1].minor = yylhsminor = self.branch_28(...);
       }
       // ...
      }
      // ...
    }

    #[inline]
    fn branch_28(&mut self, ...) {
       let col = self.yy_move(-1).yy444();
       let cd = ColumnDefinition{ col_name: col.0, col_type: col.1, constraints: self.yy_move(0).yy305() };
       let mut map = IndexMap::new();
       ColumnDefinition::add_column(&mut map, cd)?;
       return  YYMINORTYPE::yy285( map);
   }
}
```

> Do you forsee https://github.com/gwenn/lemon-rs/issues/8 to be enough or just a piece in the puzzle?

To be more explicit, what I meant is that even getting slightly below 1MB isn't great since the stack is a shared resource. So if my bigger program already uses a few hundred KB, it would still be great for trivial parse to work. Less is always better. That said, reducing the enum size would be nice because it also helps release builds despite not being a pressing issue.

Also there's clearly limits. I don't know what you had in mind. At the end of the day, the enum, i.e. tagged union, is always as big as its biggest member. We could probably squeeze some bytes with alignment/packing but the biggest wins would certainly be to move data on the heap, which would in tern come with some performance cost.

>  And do you know how to monitor the max stack size used by one program (maybe dtrace) ?

I was just looking. In principle you shouldn't need tracing or other runtime introspection, since the stack layout is determined at compile-time.

Sadly there doesn't seem to be any official tooling. And all the community tools depend on unstable APIs and eventually throw in the towel. I tried https://github.com/japaric/cargo-call-stack but couldn't get it to work.
Alternatively, one should be able to pass "-Cllvm-args=--stack-size-section" but haven't tried to extract the object information yet.

For runtime introspection you should also be able to just substract the address of a parent frame variable from the address of a child frame variable. To get the stach size of a specific object you can `use size_of::<T>()`.

**Comment by gwenn** [comment]
> We could probably squeeze some bytes with alignment/packing but the biggest wins would certainly be to move data on the heap, which would in turn come with some performance cost.

We already need to box recursive type like `Expr` so that should not be a problem.

**Comment by ignatz** [comment]
Then https://github.com/gwenn/lemon-rs/pull/79 should be fine.

If nothing else it let me calculate that yy_reduce is allocating space for ~470 instances of Stmt (stack-delta / (s0 - s1)). This also means that if we were to further half Stmt's size, we'd only safe an additional ~110kb or 14% of total stack size.

As opposed to pushing branches into their own stack frames, which would reduce the overall stack size to a few kB similar to release builds for arguably quite some work.

**Comment by ignatz** [comment]
Friendly ping

**Comment by gwenn** [comment]
Maybe as a workaround:
https://stackoverflow.com/questions/60751806/how-to-compile-some-dependencies-with-release
(I need time to investigate)

**Comment by ignatz** [comment]
Awesome, I think that's more than acceptable. Thanks.

Just for future reference: my PR is as minimal as possible. I simply looked at which types was the largest, thus defining the size of the enum/union and shrunk it, then did the same to the second largest, ... until I felt I had enough buffer with with respect to the 1MB Windows default stack size.

Even with a higher optimization level (-O1 is plenty) there's no guarantee that things will fit given that the stack is a shared resource so some reduction in size like my PR or something else would still be appreciated. Thanks

**Comment by ignatz** [comment]
Ok, https://github.com/trailbaseio/trailbase/commit/44e0bb6f846e6347125f679f13ee622a9baaa165 should have addressed the issue also for dev builds selectively bumping the optimization level. Let me know if it didn't fix your issue

**Comment by gwenn** [comment]
Just for reference:
https://github.com/apache/datafusion-sqlparser-rs/pull/540

**Comment by gwenn** [comment]
See #80 (only `Select` is boxed like sqlparser-rs)

**Comment by ignatz** [comment]
Why stop there? Boxing Limit is the next lowest hanging fruit giving another 20% improvement.

**Comment by gwenn** [comment]
See https://github.com/gwenn/lemon-rs/pull/81#issuecomment-2571519843

**Comment by ignatz** [comment]
Could I ask you to be a bit more verbose, your comment:

> ```
> use std::mem::size_of;
> 
> use sqlparser::ast::{Query, Statement};
> 
> fn main() {
>     println!("size_of(Statement) => {}", size_of::<Statement>());
>     println!("size_of(Query) => {}", size_of::<Query>());
> }
> ```
> 
> ```
> sqlparser-rs % cargo run --example sizes
> size_of(Statement) => 1856
> size_of(Query) => 1376
> ```

measures the stack sizes of some types in `sqlparser` and doesn't seem to have anything to do with `sqlite3_parser`.

What are you saying?

**Comment by gwenn** [comment]
I am just saying that a similar / popular library has an AST which size is twice bigger.
So I guess we don't need to bother...

**Comment by ignatz** [comment]
You're saying we don't have to do better because someone else is doing worse?

Do you have a technical concern?  The latest change is great, is reduces the stack size but the stack size is still ginormous (arguably even with my proposed 20% improvement). Now a single trivial parse (are there recursive situations?) **just** fits within the standard stack size of Windows (1MB) **assuming** that no other parent stack frame consumes significant stack space.

Stack space is a shared resource and the less each constituent consumes the better. For example, in the past I've worked with servers that would spawn thousands of threads. To avoid an excessive baseline, stack sizes would be limited to 64k because that's plenty for most things.

**Comment by gwenn** [comment]
https://stackoverflow.com/questions/70407682/how-to-match-value-of-t-in-boxt-that-is-attribute-of-a-struct
=> introducing boxes make it difficult to pattern match AST nodes...

**Comment by gwenn** [comment]
Version 0.16.0 released
